### DISCUSSION TITLE
Logging 3 Growatt inverters

### DISCUSSION THREAD

**Discussion by ZonPV** [johanmeijer/grott#112] [Q&A] [answered]
I know it is possible to log two Growatt inverters with Grott. 

But my brother decided to buy 3 inverters, 1 Fase for his different roofs.. A 3-fase would have just 2 strings. Now he has 6 strings..

But what about logging? We figured out that Influx and Grafana will be good enough. But I have no idea to log that data with a single Raspberry which runs with a SSD. That concept is reliable for years over here. 

Grott doesn't need a task scheduler (as far as I know). So I can run that application just once.. Only idea I have is three times a Docker container? 

Any suggestion is welcome. Thanks in advance.

**Comment by redbullterror** [comment]
if you use node-red you can use a flow to separate the inverters with a switch. and the serial numbers of the inverters

**Comment by ZonPV** [comment]
Thank you for your suggestion. But with what purpose should I use Node-Red? For filtering data from Influxdb I can make queries in Grafana. What sequence of applications did you have in mind? 
For me it starts with the question how to configure Grott.ini. That app should pick up data from the ShineWifi-X..

**Comment by johanmeijer** [comment, marked as answer]
@ZonPV 

If you use a shinelink with RF connections to the inverters it should not be any problem. Grott proxy will intercept the connection between the shinelink and the growatt server, read the data and sent it to Influxdb. 

If you have 3 inverters with a shinewifi or shinelan there are 3 connections/sessions between the inverters/dataloggers and the growatt server. Until for a week I thought this should not be a problem and the Grott proxy can handle multiple connection buts then I saw some problems with  multiple inverters and influxdb recording (see issue #110). I am not sure if that is a Grott problem or inlfuxdb. 

At this moment I am not able to test with multiple inverters.
I am busy with collecting some stuf for a second PV system to replace my 20+ your old panels (running next to my 3 year old Growatt set up) but that will take same time (and money). (I already have the inverter, only need a shinewifi / shinelink, panels and the mounting / connecting material ;) ). 

I am planning to do some more multi threading engineering (in combination with Grott acting as a server) in version 3.x of Grott but that may take a  while to develop. 

We can try if your configuration will work (if you are willing to do some debugging). 
I do not think we have to change anything in the grott.ini then. Use the basic configuration and define the connection with influxDB (if not used, disable MQTT interface). 

An other possibility is to use three Grott instances with each using a different port for the connection between the inverter / datalogger and Grott. You can still use one inluxdb server I think. 
The challenge will be to run 3 instances. With docker this is really easy. Just start 3 containers pointing to an other port. If you want to run them as a native linux deamon you need do do more work (create 3 services with each their own grott.ini). 

What @redbullterror  mean I think is that with Nodered is that you can use the standard MQTT interface and send the data to inlfuxdb with nodered. I use node red to put in my energy usage information into InfluxDB (and a lot of  other home automation stuff). 
Nodered is very easys to use (and handy for a lot of things) but might bring complexity is you only want to log the inverter data into influxDB. 
I am not sure this will solve the multiple connection error (if it is a Grott multi threading problem).

**Comment by ZonPV** [comment]
Thanks Johan for your explanation.
At the moment I have a proof of concept running on a Raspberry 2.. Because a PI version 4 is sold out everywhere. But it is running stable for a few days. Once in a while I change the IP-address in my ShineWifi-X to the Pi and after a few hours back to my Synology so I don't loose data.
For now I found out that it should be possible to run 3 separate Docker containers on a Raspberry with each a different port-number for the setup at my brothers house. Even uploading to PVoutput is possible. Or making own graphics with Grafana.

Last saturday we put the mounting rails at the roof for 26 panels. The panels will arrive within 2 weeks.. :-), together with the inverters.

So for now I will setup the PI with 3 different Docker containers with each 1 Grott Proxy server. The Docker container with Grott 2.7.1 on my Synology is as solid as a rock. Yesterday it was running for 47 days. So very pleased with it. Thanks again for that.

I am (a bit) familiar with Node-Red. I use it for conversion of data from the SMA-inverter to Domoticz.
So I am very interested in the flow you made for sending information to PV-output. No idea how to deliver data to that site with Node-Red. Maybe you can explain that a bit more in the Wiki at https://github.com/johanmeijer/grott/wiki/Node-Red,-PVOUTPUT.org,-Influxdb-and-grafana-examples or send me an example of that flow??

You wrote that there are some troubles with multi-threading in Grott-server. If I can help with debugging, just let me know! 
At the moment I have just one Growatt-inverter. Within a few weeks we have a site with 3 of those... I am intended to setup a VPN to my brothers house with VNC. So a nice playground.

For now I will give it a try with Docker containers to achieve my goal. 

Good luck with the replacement of the panels

**Comment by johanmeijer** [comment]
Somehow I lost the nodered example on github. I put it back: https://github.com/johanmeijer/grott/tree/2.7-(Beta)/examples/nodered

**Comment by rvbaelen** [comment]
is see this subject has been discussed a while a go.

My situation is this: i've got grott running in docker (configured in docker compose) with two invertors. so i've set the  - gmqttinverterintopic="True". I get data from one of the inverters, but not two.. these inverters are both identical and have both shinelan modules to commuicate. 

my guess was to use the - ginvtypemap= under inviroment section in the compose file... but i get an yamlsyntaxerror when inputting 
- ginvtypemap = "{'7AE6321422': 'inv1', '7AE6321412': 'inv2'}"  (i used not existing serial numbers) so i have to adjust this for compose i gues? :) but i'm at a dead end... dont know how.

So is the invtypemap a solution to my problem and if yes... how to adjust the env.

Thanks for your help guys!

**Comment by johanmeijer** [comment]
Invtypemap is the solotion if you have two inverters of a different type. You are able to specify the different invertype so by example tl3 for inv1 and sph for inv2. 

I think it will not help you with this problem.  gmqttinverterintopic="True" should give the information on the 2 inverters in separate MQTT topics (with the inverter name in it). 

Which version of the docker container do you use?

**Reply by rvbaelen** [reply]
you mean image? :) sorry for the lack of knowledge version 2.8.2 (i'm running on arm64 --> Pine A64-LTS).

**Reply by rvbaelen** [reply]
Dont know how... but now it does work. thanks for your help!
